File: apps/android/feature/profile/ui/src/main/java/com/scribblefit/feature/profile/ui/SettingsViewModel.kt
package com.scribblefit.feature.profile.ui

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.scribblefit.core.config.domain.ConfigRepository
import com.scribblefit.core.config.domain.LLMProvider
import com.scribblefit.core.config.domain.SystemConfig
import com.scribblefit.core.config.domain.ThemePreference
import com.scribblefit.core.config.domain.Weight
import com.scribblefit.feature.ai.domain.security.SecureKeyStorage
import com.scribblefit.feature.profile.domain.repository.ModelRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.SharingStarted
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.stateIn
import kotlinx.coroutines.launch
import javax.inject.Inject

data class SettingsUiState(
    val settings: SystemConfig,
    val availableModels: List<String> = emptyList(),
    val isLoadingModels: Boolean = false,
    val showApiKeyInput: Boolean = false
)

@HiltViewModel
class SettingsViewModel @Inject constructor(
    private val configRepository: ConfigRepository,
    private val modelRepository: ModelRepository,
    private val secureKeyStorage: SecureKeyStorage
) : ViewModel() {
    val uiState: StateFlow<SettingsUiState> = configRepository.config
        .map { config ->
            getStateFromConfig(
                config = config,
                apiKey = secureKeyStorage.getApiKey().orEmpty()
            )
        }.stateIn(
            scope = viewModelScope,
            started = SharingStarted.Eagerly,
            initialValue = SettingsUiState(
                settings = configRepository.config.value
            )
        )

    fun onProviderChanged(provider: LLMProvider) {
        viewModelScope.launch {
            val current = uiState.value.settings
            configRepository.updateConfig(
                current.copy(preferredLlmProvider = provider, preferredModel = null)
            )
        }
    }

    fun onModelSelected(model: String) {
        viewModelScope.launch {
            val current = uiState.value.settings
            configRepository.updateConfig(current.copy(preferredModel = model))
        }
    }

    fun onApiKeySaved(key: String) {
        viewModelScope.launch {
            secureKeyStorage.saveApiKey(key)
            configRepository.updateConfig(
                uiState.value.settings.copy(
                    preferredModel = null,
                    updatedAt = System.currentTimeMillis()
                )
            )
        }
    }

    fun onWeightUnitChanged(unit: Weight) {
        viewModelScope.launch {
            val current = uiState.value.settings
            configRepository.updateConfig(current.copy(weightUnit = unit))
        }
    }

    fun onThemeChanged(theme: ThemePreference) {
        viewModelScope.launch {
            val current = uiState.value.settings
            configRepository.updateConfig(current.copy(themePreference = theme))
        }
    }

    fun onClearDataTapped() {
        viewModelScope.launch {
            configRepository.resetConfig()
            secureKeyStorage.clearApiKey()
        }
    }

    private suspend fun getStateFromConfig(
        config: SystemConfig,
        apiKey: String
    ): SettingsUiState {
        val showInput = config.preferredLlmProvider.requiresApiKey
        val models = if (showInput) {
            runCatching {
                modelRepository.fetchModels(
                    provider = config.preferredLlmProvider,
                    apiKey = apiKey
                )
            }.getOrDefault(emptyList())
        } else {
            emptyList()
        }

        return SettingsUiState(
            settings = config,
            availableModels = models,
            showApiKeyInput = showInput && apiKey.isEmpty()
        )
    }
}

File: apps/ios/ScribbleFit/Sources/Features/Profile/UI/SettingsViewModel.swift
import Foundation
import Combine

public struct SettingsUiState: Sendable {
    public var config: SystemConfigDomain = SystemConfigDomain()
    public var availableModels: [String] = []
    public var isLoadingModels: Bool = false
    public var apiKey: String = ""
    public var showApiKeyField: Bool = false
}

@MainActor
public final class SettingsViewModel: ObservableObject {
    @Published public var uiState = SettingsUiState()

    private let configRepository: any ConfigRepository
    private let modelRepository: any ModelRepository
    private let secureKeyStorage: any SecureKeyStorage

    public init(
        configRepository: any ConfigRepository,
        modelRepository: any ModelRepository,
        secureKeyStorage: any SecureKeyStorage
    ) {
        self.configRepository = configRepository
        self.modelRepository = modelRepository
        self.secureKeyStorage = secureKeyStorage
        Task { await loadConfig() }
    }

    public func loadConfig() async {
        uiState.config = (await configRepository.getConfig()) ?? SystemConfigDomain()
        uiState.apiKey = await secureKeyStorage.getApiKey() ?? ""
        uiState.showApiKeyField = uiState.config.preferredLlmProvider != .proxy && uiState.config.preferredLlmProvider != .local
        if uiState.showApiKeyField && !uiState.apiKey.isEmpty {
            await fetchModels()
        }
    }

    public func onProviderChanged(_ provider: LLMProvider) async {
        let updated = SystemConfigDomain(
            summaryPrompt: uiState.config.summaryPrompt,
            suggestionPrompt: uiState.config.suggestionPrompt,
            insightPrompt: uiState.config.insightPrompt,
            parsePrompt: uiState.config.parsePrompt,
            preferredLlmProvider: provider,
            preferredModel: nil,
            weightUnit: uiState.config.weightUnit,
            themePreference: uiState.config.themePreference,
            updatedAt: Date()
        )
        uiState.config = updated
        uiState.availableModels = []
        uiState.showApiKeyField = provider != .proxy && provider != .local
        try? await configRepository.updateConfig(updated)
        if uiState.showApiKeyField && !uiState.apiKey.isEmpty {
            await fetchModels()
        }
    }

    public func onModelSelected(_ model: String) {
        let updated = SystemConfigDomain(
            summaryPrompt: uiState.config.summaryPrompt,
            suggestionPrompt: uiState.config.suggestionPrompt,
            insightPrompt: uiState.config.insightPrompt,
            parsePrompt: uiState.config.parsePrompt,
            preferredLlmProvider: uiState.config.preferredLlmProvider,
            preferredModel: model,
            weightUnit: uiState.config.weightUnit,
            themePreference: uiState.config.themePreference,
            updatedAt: Date()
        )
        uiState.config = updated
        Task { try? await configRepository.updateConfig(updated) }
    }

    public func onApiKeySaved(_ key: String) async {
        try? await secureKeyStorage.saveApiKey(key)
        uiState.apiKey = key
        await fetchModels()
    }

    public func onWeightUnitChanged(_ unit: Weight) async {
        let updated = SystemConfigDomain(
            summaryPrompt: uiState.config.summaryPrompt,
            suggestionPrompt: uiState.config.suggestionPrompt,
            insightPrompt: uiState.config.insightPrompt,
            parsePrompt: uiState.config.parsePrompt,
            preferredLlmProvider: uiState.config.preferredLlmProvider,
            preferredModel: uiState.config.preferredModel,
            weightUnit: unit,
            themePreference: uiState.config.themePreference,
            updatedAt: Date()
        )
        uiState.config = updated
        try? await configRepository.updateConfig(updated)
    }

    public func onThemeChanged(_ theme: ThemePreference) async {
        let updated = SystemConfigDomain(
            summaryPrompt: uiState.config.summaryPrompt,
            suggestionPrompt: uiState.config.suggestionPrompt,
            insightPrompt: uiState.config.insightPrompt,
            parsePrompt: uiState.config.parsePrompt,
            preferredLlmProvider: uiState.config.preferredLlmProvider,
            preferredModel: uiState.config.preferredModel,
            weightUnit: uiState.config.weightUnit,
            themePreference: theme,
            updatedAt: Date()
        )
        uiState.config = updated
        try? await configRepository.updateConfig(updated)
    }

    public func onClearDataTapped() async {
        await configRepository.resetConfig()
        try? await secureKeyStorage.clearApiKey()
        uiState = SettingsUiState()
    }

    public func fetchModels() async {
        let key = uiState.apiKey
        guard !key.isEmpty else { return }
        uiState.isLoadingModels = true
        uiState.availableModels = (try? await modelRepository.fetchModels(for: uiState.config.preferredLlmProvider, apiKey: key)) ?? []
        uiState.isLoadingModels = false
    }
}
